### PR TITLE
AST: Refactor removeSelfParam from a static function to a member function of TypeBase. NFC

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -3664,10 +3654,10 @@ ConformanceChecker::inferTypeWitnessesViaValueWitness(ValueDecl *req,
   }
 
   auto setup = [&]() -> std::tuple<Optional<RequirementMatch>, Type, Type> {
-    fullWitnessType = removeSelfParam(witness, fullWitnessType);
+    fullWitnessType = fullWitnessType->removeSelfParam(witness);
     return std::make_tuple(
         None,
-        removeSelfParam(req, req->getInterfaceType()),
+        req->getInterfaceType()->removeSelfParam(req),
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: If this is the only place it's called why not open-code the body here?

Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Also is req and witness always in type context? Then you could have an assert here, instead of an 'if'.

Comment (nkcsgexi) [inline lib/Sema/TypeCheckProtocol.cpp]: I plan to use this in IDE type checking APIs, more specifically, to compare the interface types of two function decls without Self context.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Maybe it could be a method on the decl then, since you're always calling it on foo->getInterfaceType()?

Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: There are other places we do stuff like getInterfaceType()->castTo<AnyFunctionType>()->getResult()... if you grep for variations on this you'll find other places to use the new utility method.

(Of course this is optional; I'm not asking you to do anything before merging the PR)

Comment (nkcsgexi) [inline lib/Sema/TypeCheckProtocol.cpp]: Another question: is interface type always canonical?

**Pull request closed without merging**